arch/arm/src/stm32wl5/stm32wl5_spi.c

```diff
@@ -0,0 +1,1858 @@
+  ret = (uint32_t)spi_readword(priv);
+
+  /* Check and clear any error flags
+   * (Reading from the SR clears the error flags)
+   */
+
+  regval = spi_getreg(priv, STM32WL5_SPI_SR_OFFSET);
+
+  spiinfo("Sent: %04" PRIx32 " Return: %04" PRIx32
+          " Status: %02" PRIx32 "\n", wd, ret, regval);
+  UNUSED(regval);
+
+  return ret;
+}
+
+/****************************************************************************
+ * Name: spi_exchange (no DMA).  aka spi_exchange_nodma
+ *
+ * Description:
+ *   Exchange a block of data on SPI without using DMA
+ *
+ *   REVISIT:
+ *   This function could be much more efficient by exploiting (1) RX and TX
+ *   FIFOs and (2) the STM32 F3 data packing.
+ *
+ * Input Parameters:
+ *   dev      - Device-specific state data
+ *   txbuffer - A pointer to the buffer of data to be sent
+ *   rxbuffer - A pointer to a buffer in which to receive data
+ *   nwords   - the length of data to be exchanged in units of words.
+ *              The wordsize is determined by the number of bits-per-word
+ *              selected for the SPI interface.  If nbits <= 8, the data is
+ *              packed into uint8_t's; if nbits >8, the data is packed into
+ *              uint16_t's
+ *
+ * Returned Value:
+ *   None
+ *
+ ****************************************************************************/
+
+#if !defined(CONFIG_STM32WL5_SPI_DMA) || defined(CONFIG_STM32WL5_DMACAPABLE) || \
+     defined(CONFIG_STM32WL5_SPI_DMATHRESHOLD)
+#if !defined(CONFIG_STM32WL5_SPI_DMA)
+static void spi_exchange(struct spi_dev_s *dev, const void *txbuffer,
+                         void *rxbuffer, size_t nwords)
+#else
+static void spi_exchange_nodma(struct spi_dev_s *dev,
+                               const void *txbuffer,
+                               void *rxbuffer, size_t nwords)
+#endif
+{
+  struct stm32wl5_spidev_s *priv = (struct stm32wl5_spidev_s *)dev;
+  DEBUGASSERT(priv && priv->spibase);
+
+  spiinfo("txbuffer=%p rxbuffer=%p nwords=%d\n", txbuffer, rxbuffer, nwords);
+
+  /* 8- or 16-bit mode? */
+
+  if (priv->nbits > 8)
+    {
+      /* 16-bit mode */
+
+      const uint16_t *src  = (const uint16_t *)txbuffer;
+            uint16_t *dest = (uint16_t *)rxbuffer;
+            uint16_t  word;
+
+      while (nwords-- > 0)
+        {
+          /* Get the next word to write.  Is there a source buffer? */
+
+          if (src)
+            {
+              word = *src++;
+            }
+          else
+            {
+              word = 0xffff;
+            }
+
+          /* Exchange one word */
+
+          word = (uint16_t)spi_send(dev, (uint32_t)word);
+
+          /* Is there a buffer to receive the return value? */
+
+          if (dest)
+            {
+              *dest++ = word;
+            }
+        }
+    }
+  else
+    {
+      /* 8-bit mode */
+
+      const uint8_t *src  = (const uint8_t *)txbuffer;
+            uint8_t *dest = (uint8_t *)rxbuffer;
+            uint8_t  word;
```

Comment:
```suggestion
      const uint8_t *src  = (const uint8_t *)txbuffer;
      uint8_t *dest = (uint8_t *)rxbuffer;
      uint8_t  word;
```

Reply:
It is fixed. The question is if keep or remove DMA code. Those code base on stm32 SPI driver. I suppose that there are bugs in DMA for stm32 SPI driver.

What to do. Keep or remove DMA, not fully implemented code?

Reply:
I'm not sure. You can leave the code and add `#error` if DMA is enabled with a proper message. Or remove DMA